### PR TITLE
Improve messaging

### LINKED ISSUE #3: Need the ability to create a safe if it does not exist with the correct settings

### PR DESCRIPTION
**Pull request by hjbotha-cybr**
Implemented Write-LogMessage function
Moved most messages to Verbose channel
Logs will be saved in _Set-DomainUser.log
PSM safe will be created if it does not exist and appropriate permissions assigned
If login to PVWA fails, a useful error will be displayed, whether it's a connection error or a login failure

### CODE REVIEW

File: Set-DomainUser.ps1 [hunk 1]
```diff
@@ -1161,8 +1257,9 @@ Write-LogMessage -Type Verbose -MSG "Checking current safe status"
 $safeStatus = Get-SafeStatus -pvwaAddress $pvwaAddress -pvwaToken $pvwaToken -SafeName $Safe
 if ($safeStatus -eq $false) {
     # function returns false if safe does not exist
-    Write-LogMessage -Type Error -MSG "Safe $Safe does not exist. Please create it or provide a different safe name with the -Safe option"
-    exit 1
+    Write-LogMessage -Type Error -MSG "Safe $Safe does not exist. Creating the safe now"
```
Comment (hjbotha-cybr) [inline Set-DomainUser.ps1:1260]: Should be type Verbose or Info

File: Set-DomainUser.ps1 [hunk 2]
```diff
@@ -303,18 +445,20 @@ Function Test-PvwaToken {
...
     $testToken = Invoke-RestMethod -Method 'Get' -Uri $url -Headers $Headers -ContentType 'application/json'
     if ($testToken) {
-        return $true
+        return @{
+            ErrorCode = "Success"
+        }
     }
-    else {
-        return $false
+    Catch {
```
Comment (Ethlas) [inline Set-DomainUser.ps1:462]: We do not have a try part for this message.

File: Set-DomainUser.ps1 [hunk 3]
```diff
@@ -752,12 +896,173 @@ Function Activate-Platform {
...
+            viewAuditLog                           = $False
+            viewSafeMembers                        = $False
+            accessWithoutConfirmation              = $False
+            createFolders                          = $False
+            deleteFolders                          = $False
+            moveAccountsAndFolders                 = $False
+            requestsAuthorizationLevel1            = $False 
+            requestsAuthorizationLevel2            = $False
+        }
+        Set-SafePermissions -pvwaAddress $pvwaAddress -pvwaToken $pvwaToken -safe $safe -safeMember "PSMAppUsers" -safePermissions $PSMAppUsers
```
Comment (Ethlas) [inline Set-DomainUser.ps1:993]: Not sure why this is set and not new. I think it should also be new.

File: Set-DomainUser.ps1 [hunk 4]
```diff
@@ -752,12 +896,173 @@ Function Activate-Platform {
...
+        Set-SafePermissions -pvwaAddress $pvwaAddress -pvwaToken $pvwaToken -safe $safe -safeMember "PSMAppUsers" -safePermissions $PSMAppUsers  
+        return $true      
+    }
+    catch {
+        Write-LogMessage -Type Error $_.ErrorDetails.Message
+        return $false
+    }
+}
+
+Function Set-SafePermissions {
```
Comment (Ethlas) [inline Set-DomainUser.ps1:1002]: Why did I make a set and a new function? I think set should only be if the permissions are already there?

File: Set-DomainUser.ps1 [hunk 5]
```diff
@@ -971,375 +1274,234 @@ if (!($NETBIOS)) {
...
+if (!($pvwaTokenResponse.Response -match "[0-9a-zA-Z]{200,256}")) {
+    # If we get here, it means we got a 200 response from the server, but the data it returned was not a valid token.
+    # In this case, we display the response we got from the server to aid troubleshooting.
+    Write-LogMessage -Type Error "Response from server was not a valid token:"
+    Write-LogMessage -Type Error $pvwaTokenResponse.Response
+    exit 1
+}
+# If we get here, the token was retrieved successfully and looks valid. We'll still test it though.
+$PvwaTokenTestResponse = Test-PvwaToken -Token $pvwaTokenResponse.Response -pvwaAddress $pvwaAddress
+if ($PvwaTokenTestResponse.ErrorCode -eq "Success") {
```
Comment (Ethlas) [inline Set-DomainUser.ps1:1335]: Should we use an actual code rather than a string?

**Review by Ethlas — COMMENTED** [review]
Put a few notes. A couple are for me if you can also review my stuff so we can merge it.

File: Set-DomainUser.ps1 (hunk 3 above)
Comment (hjbotha-cybr) [reply Set-DomainUser.ps1:993]: I wonder if there's an environmental difference. In mine, PSMAppUsers was added to the safe automatically at creation time and trying to add it again resulted in an error, so I added the Set function to modify the existing user. Is the behaviour different in yours? May need to add something to check if the member exists and use Set or New depending.

File: Set-DomainUser.ps1 (hunk 4 above)
Comment (hjbotha-cybr) [reply Set-DomainUser.ps1:1002]: I added the new function

**Comment by hjbotha-cybr** [comment]
> We do not have a try part for this message.  

fixed

File: Set-DomainUser.ps1 (hunk 5 above)
Comment (hjbotha-cybr) [reply Set-DomainUser.ps1:1335]: We can adjust if we feel the need. Doesn't really matter what we use as long as it's unambiguous and unique. It's never shown to the user, just used to check the result of the function.

**Pull request merged**